nova/scheduler/weights/ram.py

```diff
@@ -36,3 +36,24 @@ def weight_multiplier(self):
     def _weigh_object(self, host_state, weight_properties):
         """Higher weights win.  We want spreading to be the default."""
         return host_state.free_ram_mb
+
+
+class RamResizeBigVmToSameHostWeigher(weights.BaseHostWeigher):
+    minval = 0
+
+    def weight_multiplier(self):
+        """Override the weight multiplier."""
+        return CONF.filter_scheduler.ram_resize_to_same_host_weight_multiplier
+
+    def _weigh_object(self, host_state, weight_properties):
+        """Return 1 for big VMs when the host is the instances current host.
+        Return 0 otherwise.
+        """
+        instance_uuid = weight_properties.instance_uuid
+        if instance_uuid in host_state.instances:
```

Comment:
Imho the function should just contain a check
```python
if instance_uuid not in host_state.instance:
  return 0.0

if not is_a_resize():
  return 0.0

return 1.0
```
With `is_a_resize()` being the hard part. Maybe you can check if there's a flavor change going on and/or the task state or instance state.